kong/clustering/rpc/socket.lua

```diff
@@ -154,54 +253,52 @@ function _M:start()
       assert(typ == "binary")
 
       local payload = decompress_payload(data)
-      assert(payload.jsonrpc == jsonrpc.VERSION)
-
-      if payload.method then
-        -- invoke
-
-        ngx_log(ngx_DEBUG, "[rpc] got RPC call: ", payload.method, " (id: ", payload.id, ")")
 
-        local dispatch_cb = self.manager.callbacks.callbacks[payload.method]
-        if not dispatch_cb and payload.id then
-          local res, err = self.outgoing:push(new_error(payload.id, jsonrpc.METHOD_NOT_FOUND))
-          if not res then
-            return nil, "unable to send \"METHOD_NOT_FOUND\" error back to client: " .. err
-          end
-
-          goto continue
+      -- single rpc call
+      if not isarray(payload) then
+        local ok, err = self:process_rpc_msg(payload)
+        if not ok then
+          return nil, err
         end
 
-        -- call dispatch
-        local res, err = kong.timer:named_at(string_format("JSON-RPC callback for node_id: %s, id: %d, method: %s",
-                                                           self.node_id, payload.id or 0, payload.method),
-                                                           0, _M._dispatch, self, dispatch_cb, payload)
-        if not res and payload.id then
-          local reso, erro = self.outgoing:push(new_error(payload.id, jsonrpc.INTERNAL_ERROR))
-          if not reso then
-            return nil, "unable to send \"INTERNAL_ERROR\" error back to client: " .. erro
-          end
+        goto continue
+      end
 
-          return nil, "unable to dispatch JSON-RPC callback: " .. err
+      -- rpc call with an empty Array
+      if isempty(payload) then
+        local res, err = self:push_response(
+                          new_error(nil, jsonrpc.INVALID_REQUEST, "empty batch array"))
+        if not res then
+          return nil, err
         end
 
-      else
-        -- response
-        local interest_cb = self.interest[payload.id]
-        self.interest[payload.id] = nil -- edge trigger only once
+        goto continue
+      end
 
-        if not interest_cb then
-          ngx_log(ngx_WARN, "[rpc] no interest for RPC response id: ", payload.id, ", dropping it")
+      -- batch rpc call
 
-          goto continue
-        end
+      local collection = {}
 
-        local res, err = interest_cb(payload)
-        if not res then
-          ngx_log(ngx_WARN, "[rpc] RPC response interest handler failed: id: ",
-                  payload.id, ", err: ", err)
+      for _, v in ipairs(payload) do
+        local ok, err = self:process_rpc_msg(v, collection)
+        if not ok then
+          return nil, err
         end
       end
```

Comment:
For batch all, the `self:process_rpc_msg` will process each request by spawning a new timer. Is this expected? Shouldn't we wait for all spawning timers to exit?